linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala

```diff
@@ -0,0 +1,406 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.backend.webassembly
+
+import scala.collection.mutable
+
+import org.scalajs.ir.{OriginalName, Position}
+
+import Instructions._
+import Identitities._
+import Modules._
+import Types._
+
+final class FunctionBuilder(
+    moduleBuilder: ModuleBuilder,
+    val functionID: FunctionID,
+    val functionOriginalName: OriginalName,
+    functionPos: Position
+) {
+  import FunctionBuilder._
+
+  private var labelIdx = 0
+
+  private val params = mutable.ListBuffer.empty[Local]
+  private val locals = mutable.ListBuffer.empty[Local]
+  private var resultTypes: List[Type] = Nil
+
+  private var specialFunctionType: Option[TypeID] = None
+
+  /** The instructions buffer. */
+  private val instrs: mutable.ListBuffer[Instr] = mutable.ListBuffer.empty
+
+  def setFunctionType(typeID: TypeID): Unit =
+    specialFunctionType = Some(typeID)
+
+  def setResultTypes(tpes: List[Type]): Unit =
+    resultTypes = tpes
+
+  def setResultType(tpe: Type): Unit =
+    setResultTypes(tpe :: Nil)
+
+  def addParam(originalName: OriginalName, tpe: Type): LocalID = {
+    val id = new ParamIDImpl(params.size, originalName)
+    params += Local(id, originalName, tpe)
+    id
+  }
+
+  def addParam(name: String, tpe: Type): LocalID =
+    addParam(OriginalName(name), tpe)
+
+  def genLabel(): LabelID = {
+    val label = new LabelIDImpl(labelIdx)
+    labelIdx += 1
+    label
+  }
+
+  def addLocal(originalName: OriginalName, tpe: Type): LocalID = {
+    val id = new LocalIDImpl(locals.size, originalName)
+    locals += Local(id, originalName, tpe)
+    id
+  }
+
+  def addLocal(name: String, tpe: Type): LocalID =
+    addLocal(OriginalName(name), tpe)
+
+  // Instructions
+
+  def +=(instr: Instr): Unit =
+    instrs += instr
+
+  def ++=(instrs: Iterable[Instr]): Unit =
+    this.instrs ++= instrs
+
+  def markCurrentInstructionIndex(): InstructionIndex =
+    new InstructionIndex(instrs.size)
+
+  def insert(index: InstructionIndex, instr: Instr): Unit =
+    instrs.insert(index.value, instr)
```

Comment:
We do create multiple marks that are live at the same time. What matters is that they're created and used in a stack order. For example "create 1, create 2, use 2, use 1" is valid, but "create 1, create2, use1, use 2" is not.

Reply:
It looks like this is only ever used to insert "conditional blocks". Do you think we could amend the API to something like (maybe with additional params for the block type):

```
def maybeBlock(body: => Unit)(cond: => Boolean)(tail: => Unit): Unit
```

IIUC this would:
- Cover existing usage (without really impeding control flow at the usage site).
- Enforce that marks are used in the right order.

Reply:
That doesn't work as nicely as that. If you look at the call site in `genLabeled`, there are two things that make this problematic:
* The `label` of the optional `Block` can only be obtained if the `cond` succeeds (because of the call to `requireCrossInfo()`).
* There are conditional instructions even *after* the optional block's `end`.

So the signature would have to be
```scala
def maybeBlock(body: => Unit)(cond: => Boolean)(blockLabel: Option[LabelID])(
    tailInsideBlock: =>Unit)(tailAfterBlock: => Unit): Unit
```
which becomes super specialized to its call site, at that point.

Reply:
Fair enough. I guess we'll have to live with it :)